include/onnxruntime/core/framework/allocator.h

```diff
@@ -15,21 +15,78 @@
 #include "core/framework/fence.h"
 #include "core/session/onnxruntime_c_api.h"
 
+// Struct to represent a physical device.
+struct OrtDevice {
+  typedef int8_t DeviceType;
+  typedef int8_t MemoryType;
+  typedef int16_t DeviceId;
+
+  // Pre-defined device types.
+  static const DeviceType CPU = 0;
+  static const DeviceType GPU = 1; //CUDA
+  static const DeviceType FPGA = 2;
+
+  // Pre-defined memory types.
+  static const MemoryType Default = 0;
+  static const MemoryType CUDA_PINNED = 1;
+
+  constexpr OrtDevice(DeviceType device_type_, MemoryType memory_type_, DeviceId device_id_)
+      : device_type(device_type_),
+        memory_type(memory_type_),
+        device_id(device_id_) {}
+
+  constexpr OrtDevice() : OrtDevice(CPU, Default, 0) {}
+
+  DeviceType Type() const {
+    return device_type;
+  }
+
+  MemoryType MemType() const {
+    return memory_type;
+  }
+
+  DeviceId Id() const {
+    return device_id;
+  }
+
+  inline std::string ToString() const {
+    std::ostringstream ostr;
+    ostr << "Device: ["
+         << " type:" << device_type
+         << " memory_type:" << memory_type
+         << " device_id:" << device_id
+         << "]";
+    return ostr.str();
+  }
+
+ private:
+  // Device type.
+  DeviceType device_type;
+
+  // Memory type.
+  MemoryType memory_type;
+
+  // Device index.
+  DeviceId device_id;
+};
+
 struct OrtAllocatorInfo {
   // use string for name, so we could have customized allocator in execution provider.
   const char* name;
   int id;
   OrtMemType mem_type;
   OrtAllocatorType type;
+  OrtDevice device;
```

Comment:
Is the memory type in OrtDevice same as OrtMemType? Looks like this info is getting duplicated here.

Reply:
The design should be the same. However, the bad thing is the OrtMemType was ruined. It mixes tensor property with mem type together. For example, OrtMemTypeCPUOutput/OrtMemTypeCPU is actually CPU memory, but Cuda_Pinned. OrtMemTypeCPUInput and default in CPU execution provider case should be also the same.

I'm introducing this new memory type in OrtDevice, and will refactor/remove the OrtMemType later in separate PRs.
Thank you!